optuna/_gp/batched_lbfgsb.py

```diff
@@ -70,30 +70,31 @@ def _func_and_grad(x: np.ndarray) -> tuple[float, np.ndarray]:
         xs_opt[i] = x_opt
         fvals_opt[i] = fval_opt
         n_iterations[i] = info["nit"]
-        is_remaining_batch[i] = False
 
     greenlets = [greenlet(run) for _ in range(batch_size)]
-    x_batched = [gl.switch(i) for i, gl in enumerate(greenlets)]
-
-    while np.any(is_remaining_batch):
-        remaining_batch_indices = np.where(is_remaining_batch)[0]
-        fvals, grads = func_and_grad(np.asarray(x_batched), np.asarray(remaining_batch_indices))
-
-        x_batched = []
-        next_greenlets = []
-        for i, gl in enumerate(greenlets):
-            x = gl.switch((fvals[i], grads[i]))
-            if x is None:
-                continue
-            x_batched.append(x)
-            next_greenlets.append(gl)
-        greenlets = next_greenlets
+    x_and_argsT_pairs = [gl.switch(i) for i, gl in enumerate(greenlets)]
+
+    while x_and_argsT_pairs:
+        x_batched_list = [pair[0] for pair in x_and_argsT_pairs if pair is not None]
+        args_tuple_transposed = [pair[1] for pair in x_and_argsT_pairs if pair is not None]
+        x_batched = np.array(x_batched_list)
+        current_args_tuple = tuple(zip(*args_tuple_transposed))
+        fvals, grads = func_and_grad(x_batched, *current_args_tuple)
+        results = [gl.switch((fvals[i], grads[i])) for i, gl in enumerate(greenlets)]
+        live_pairs_and_greenlets = [
+            (pair, gl) for pair, gl in zip(results, greenlets) if pair is not None
+        ]
+        if not live_pairs_and_greenlets:
+            break
+        x_and_argsT_pairs, greenlets = map(list, zip(*live_pairs_and_greenlets))
+
```

Comment:
Much shorter, but the identical behavior with the following.

> [!NOTE]
> You need to modify the last line of `run`, which corresponds to `greenlet.getcurrent().parent.switch(None, None)`.

```suggestion
        greenlet.getcurrent().parent.switch(None, None)

    greenlets = [greenlet(run) for _ in range(batch_size)]
    x_and_args_list = [gl.switch(i) for i, gl in enumerate(greenlets)]
    batched_args = tuple(zip(*[args for _, args in x_and_args_list]))
    while (x_batched := np.array([x for x, _ in x_and_args_list if x is not None])).size:
        fvals, grads = func_and_grad(x_batched, *batched_args)
        x_and_args_list = [gl.switch(fvals[i], grads[i]) for i, gl in enumerate(greenlets)]
        greenlets = [gl for (x, _), gl in zip(x_and_args_list, greenlets) if x is not None]
        batched_args = tuple(zip(*[args for _, args in x_and_args_list if args is not None]))
```

Reply:
Thanks for your suggestion! This looks much clearer, so I will modify it.
In my opinion, the code below is almost identical to yours, but it looks simpler because the same processing is repeated in the same order. (However, it might appear more verbose.)
I have already committed this change.
```
...
greenlets = [greenlet(run) for _ in range(batch_size)]
x_and_args_list = [gl.switch(i) for i, gl in enumerate(greenlets)]
x_batched = np.array([x for x, _ in x_and_args_list if x is not None])
batched_args = tuple(zip(*[args for _, args in x_and_args_list if args is not None]))
while x_batched.size:
    fvals, grads = func_and_grad(x_batched, *batched_args)
    x_and_args_list = [gl.switch(fvals[i], grads[i]) for i, gl in enumerate(greenlets)]
    x_batched = np.array([x for x, _ in x_and_args_list if x is not None])
    batched_args = tuple(zip(*[args for _, args in x_and_args_list if args is not None]))
    greenlets = [gl for (x, _), gl in zip(x_and_args_list, greenlets) if x is not None]
...
```